src/game.cpp

```diff
@@ -3196,6 +3140,84 @@ bool game::handle_action()
     return (!u.is_dead_state());
 }
 
+bool game::try_get_left_click_action( action_id &act, const tripoint &mouse_target ) {
+    bool new_destination = true;
+    if (!destination_preview.empty()) {
+        auto &final_destination = destination_preview.back();
+        if (final_destination.x == mouse_target.x && final_destination.y == mouse_target.y) {
+            // Second click
+            new_destination = false;
+            u.set_destination(destination_preview);
+            destination_preview.clear();
+            act = u.get_next_auto_move_direction();
+            if (act == ACTION_NULL) {
+                // Something went wrong
+                u.clear_destination();
+                return false;
+            }
+        }
+    }
+
+    if (new_destination) {
+        destination_preview = m.route( u.pos(), mouse_target, 0, 1000 );
+        return false;
+    }
+
+    return true;
+}
+
+bool game::try_get_right_click_action( action_id &act, const tripoint &mouse_target ) {
+    const bool cleared_destination = !destination_preview.empty();
+    u.clear_destination();
+    destination_preview.clear();
+
+    if (cleared_destination) {
+        // Produce no-op if auto-move had just been cleared on this action
+        // e.g. from a previous single left mouse click. This has the effect
+        // of right-click cancelling an auto-move before it is initiated.
+        return false;
+    }
+
+    int mouse_selected_mondex = mon_at( mouse_target );
+    if (mouse_selected_mondex != -1) {
+        monster &critter = critter_tracker->find(mouse_selected_mondex);
+        if (!u.sees(critter)) {
+            add_msg(_("Nothing relevant here."));
+            return false;
+        }
+
+        if (!u.weapon.is_gun()) {
+            add_msg(m_info, _("You are not wielding a ranged weapon."));
+            return false;
+        }
+
+        //TODO: Add weapon range check. This requires weapon to be reloaded.
+
+        act = ACTION_FIRE;
+    } else if (std::abs(mouse_target.x - u.posx()) <= 1 && std::abs(mouse_target.y - u.posy()) <= 1 &&
+            m.close_door( tripoint( mouse_target.x, mouse_target.y, u.posz() ), !m.is_outside(u.pos()), true)) {
+        // Can only close doors when adjacent to it.
+        act = ACTION_CLOSE;
+    } else {
+        int dx = abs(u.posx() - mouse_target.x);
+        int dy = abs(u.posy() - mouse_target.y);
+        if (dx < 2 && dy < 2) {
+            if (dy == 0 && dx == 0) {
+                // Clicked on self
+                act = ACTION_PICKUP;
+            } else {
+                // Clicked adjacent tile
```

Comment:
The "is adjacent" check is duplicated a few lines above in the condition for `ACTION_CLOSE`. It would be nice to extract that check into a boolean constant at the top of the function and use that in both conditions.

This may also allow you to get rid of the redundant comments.